File: Medium/Maximum Product of Splitted Binary Tree/maximum-product-of-splitted-binary-tree.py
class Solution:
    def maxProduct(self, root: Optional[TreeNode]) -> int:

        total = 0
        sub_totals = []

        def search(node):
            nonlocal total
            if node is None:
                return 0
            
            total += node.val

            subtree_val = node.val
            subtree_val += search(node.left)
            subtree_val += search(node.right)
            sub_totals.append(subtree_val)

            return subtree_val
        
        search(root)
        res = -inf

        for st in sub_totals:
            res = max(res, (total-st) * st)

        return res % (10**9 + 7)

File: Medium/Maximum Product of Splitted Binary Tree/maximum-product-of-splitted-binary-tree.cpp
class Solution {
public:
    int maxProduct(TreeNode* root) {

        long total = 0;
        std::vector<long> subtrees;

        std::function<long(TreeNode*)> search = [&](TreeNode* node)
        {
            if (node == nullptr) return static_cast<long>(0);

            total += node->val;

            long subtotal = node->val;
            subtotal += search(node->left);
            subtotal += search(node->right);
            subtrees.emplace_back(subtotal);

            return subtotal;
        };

        search(root);

        long max = 0;
        for (auto st : subtrees)
            max = std::max(max, (total-st) * st );

        return max % static_cast<long>(1e9+7);
    }
};
